### ISSUE TITLE
Be able to deliberately ignore fields?

### ISSUE THREAD

**Issue by michaeljones** [zwilias/json-decode-exploration#1]
I think it might be helpful to have an 'ignore' function in the Pipeline API and maybe the simple one. Not 100% sure what it would look like but sometimes there are fields that I know are there but I don't want to decode and I don't want warnings about them either. 

I could check the warnings and make sure they are what I expect but it seems like it would be easier to declare what fields I'd like to ignore in the decoders themselves:

```elm
decode MyData
  |> required "id" int
  |> required "value" bool
  |> ignore "type"
```

Does that seem reasonable? Or is there perhaps an easy way to achieve it already?

---

It seems this does the trick:

```elm
ignore : String -> Decode.Decoder a -> Decode.Decoder b -> Decode.Decoder b
ignore key valueDecoder decoder =
    Decode.field key valueDecoder
        |> Decode.andThen (\_ -> decoder)


ignoreAt : List String -> Decode.Decoder a -> Decode.Decoder b -> Decode.Decoder b
ignoreAt path valueDecoder decoder =
    Decode.at path valueDecoder
        |> Decode.andThen (\_ -> decoder)
```
I guess it is up to your opinion on API design. I think this might be quite useful to have but as it is easily achievable with the current tools then maybe there is a argument that it is redundant.

Though they required you to specify the decoder for the field value, though that could be avoided by using the value decoder I guess. Probably side on having more type safety than less!

**Comment by zwilias** [comment]
`value` recursively marks things as "used", so indeed, that can be leveraged to ignore entire subtrees.

I like having a pipeline-specific API for it, that makes sense to me. 👍 

One question is whether using `ignore` on a non-existing key should show up as a warning or an error 🤔  I'm going to think on this some more before implementing anything.

**Comment by michaeljones** [comment]
Sounds good, always worth taking some time. 

I've decided to push it a little further on my end. The pure ignore was making me uneasy so I've implemented a kind of checkedIgnore:

```elm
checkedIgnore : String -> Decode.Decoder a -> a -> Decode.Decoder b -> Decode.Decoder b
checkedIgnore key valueDecoder value decoder =
    Decode.field key valueDecoder
        |> Decode.andThen
            (\decodedValue ->
                if decodedValue == value then
                    decoder
                else
                    Decode.fail
                        ("Unexpected value in key '"
                            ++ key
                            ++ "'. Expected: "
                            ++ Basics.toString value
                            ++ ". Actual: "
                            ++ Basics.toString decodedValue
                        )
            )


checkedIgnoreAt : List String -> Decode.Decoder a -> a -> Decode.Decoder b -> Decode.Decoder b
checkedIgnoreAt path valueDecoder value decoder =
    Decode.at path valueDecoder
        |> Decode.andThen
            (\decodedValue ->
                if decodedValue == value then
                    decoder
                else
                    Decode.fail
                        ("Unexpected value in path '"
                            ++ String.join ", " path
                            ++ "'. Expected: "
                            ++ Basics.toString value
                            ++ ". Actual: "
                            ++ Basics.toString decodedValue
                        )
            )
```

With some tests:

```elm
all : Test
all =
    describe "Zap.Json.Decode.Explorative.Pipeline"
        [ describe "checkedIgnore"
            [ test "should succeed if matches" <|
                \() ->
                    let
                        decoder =
                            decode (,)
                                |> required "a" int
                                |> required "b" int
                                |> checkedIgnore "c" int 3
                    in
                    """{"a": 1, "b": 2, "c": 3}"""
                        |> Decode.decodeString decoder
                        |> Expect.equal (Success ( 1, 2 ))
            , test "should fail if it doesn't match" <|
                \() ->
                    let
                        decoder =
                            decode (,)
                                |> required "a" int
                                |> required "b" int
                                |> checkedIgnore "c" int 2

                        jsonString =
                            """{"a": 1, "b": 2, "c": 3}"""

                        jsonData =
                            case Json.Decode.decodeString Json.Decode.value jsonString of
                                Ok data ->
                                    data

                                Err _ ->
                                    Debug.crash "Failed to decode string"
                    in
                    jsonString
                        |> Decode.decodeString decoder
                        |> Expect.equal
                            (Errors
                                (Nonempty
                                    (Decode.Failure "Unexpected value in key 'c'. Expected: 2. Actual: 3" jsonData)
                                    []
                                )
                            )
            , test "should fail if missing" <|
                \() ->
                    let
                        decoder =
                            decode (,)
                                |> required "a" int
                                |> required "b" int
                                |> checkedIgnore "c" int 2

                        jsonString =
                            """{"a": 1, "b": 2}"""

                        jsonData =
                            case Json.Decode.decodeString Json.Decode.value jsonString of
                                Ok data ->
                                    data

                                Err _ ->
                                    Debug.crash "Failed to decode string"
                    in
                    jsonString
                        |> Decode.decodeString decoder
                        |> Expect.equal
                            (Errors
                                (Nonempty
                                    (Decode.Failure "Expected an object with a field 'c'" jsonData)
                                    []
                                )
                            )
            ]
        ]
```
Usage:

```elm
decode MyData
  |> required "id" int
  |> required "text" string
  |> checkedIgnore "type" string "question"
```

I'm sure there is much that can be improved but you get the idea.

**Comment by zwilias** [comment]
This, I like, and it fits with an idea that I've been wanting to implement in a library:

`Json.Decode.check : Decoder a -> a -> Decoder a` which would succeed iff the decoded value matches the expected value.

Given a type `type Pet = Cat | Dog`, this could be used like so:
```elm
petDecoder : Decoder Pet
petDecoder =
    oneOf
        [ check string "cat" |> map (always Cat)
        , check string "dog" |> map (always Dog)
        ]

""" [ "cat", "dog", "dog", "cat" ] """
    |> decodeString petDecoder
--> Success [ Cat, Dog, Dog, Cat ]
```

Adding a nice API for pipeline (I like `checked` and `checkedAt` personally), would be pretty simple, too. Thanks, this is great input!

**Comment by michaeljones** [comment]
Sounds great! That API looks very nice. I don't enjoy the verbosity of dealing with simple 'string' -> 'union type' decoding at the moment.

I'd almost be tempted to implement a `ifCheck` or `checkThen` or something which would shortcut the need for the `|> map (always` but it does start to be diminishing returns at some point. 

Thanks for the quick responses. Happy to be helping. This library is proving very useful for us so I'm grateful.

**Comment by zwilias** [comment]
Hm, fair point, a `check : Decoder a -> a -> Decoder b` could also work; so one can `succeed Cat`, or `fancyCatDecoder`, etc.

**Comment by zwilias** [comment]
Published as 4.2.0

**Comment by zoechi** [comment]
When a field is not parsed, I get a warning.
When I `ignored` it and it does not exist I get an error.
Is there a way to not be bothered at all about a field whether it exists or not?
